### ISSUE TITLE
Metas with reference cells

### ISSUE THREAD

**Issue by atennapel** [AndrasKovacs/elaboration-zoo#8]
In an efficient implementation would you use reference cells for metas? Or does that cause other problems?

When implementing typeclasses I think using reference cells is an issue with NbE because you have to quote the whole type in order to refresh the metas. Unless there is another way?

**Comment by AndrasKovacs** [comment]
I used a [mutable array](https://github.com/AndrasKovacs/smalltt/blob/master/src/ElabState.hs#L50) for metas in efficient implementation. You can use mutable references as well; in Haskell specifically the former is faster as far as I know. Typeclasses shouldn't change anything with respect to updating metas. In all of my implementations I  do not quote whole types in order to update metas, I instead force values until a "weak head normal" form before doing pattern matching on them.

Also, even in an efficient implementation, storing metas in a persistent structure can be practical if we intend to frequently backtrack on solutions. If meta update is destructive, then rolling back updates is harder.

**Comment by atennapel** [comment]
Thanks! With typeclasses I mean that you have to match types with instances, which will mutate the meta array. So before matching a type I guess you would have backup the meta array so you can backtrack if the instance is not a match. Does that sound right to you?

**Comment by AndrasKovacs** [comment]
@atennapel I would keep metas used in instance head matching separate from normal metas, because no normal metas should be solved during instance resolution. Metas used in instance matching are temporary; they can be deleted after a failed or successful resolution and they never appear in elaboration output. 

However, they can persist for a while if instance resolution is postponed because of a normal meta. For example, if we want to resolve `Eq α` where `α` is a regular meta, then we have to postpone resolution until the meta is solved.

**Comment by atennapel** [comment]
If I understand correctly, then you can postpone the typeclass constraints (such as `Eq ?t`) until all the metas from the type inference are solved? Meaning that at the end the only metas are the ones that are created from the instance heads. With backtracking I meant more that when the instance head match fails  we have to revert the meta solution, but I see now that if there are no unsolved metas in the typeclass constraint this will not be necessary.

**Comment by AndrasKovacs** [comment]
I'm happy to answer any question here.

> If I understand correctly, then you can postpone the typeclass constraints (such as Eq ?t) until all the metas from the type inference are solved? 

We postpone Eq ?t only until ?t is solved, then we retry resolution. At that time, it's possible that we are again blocked by some ?u, in which case we postpone again.

> With backtracking I meant more that when the instance head match fails we have to revert the meta solution, but I see now that if there are no unsolved metas in the typeclass constraint this will not be necessary.

Let me expand a bit more on this.

Instance resolution can succeed even if a class constraint has metas. For example:

    class Foo a
    instance Foo a

Now, `Foo ?t` is solved because we have instance for any type. What can *not* do, is that we solve an elaboration meta based on instance resolution.

    class Foo a
    instance Foo Bool
    instance Foo Nat

Then, if we have `Foo ?t`, we can **not** solve ?t to Bool while matching the instance head. It should be clear that that would be an extremely weird behavior. Instead, we postpone `Foo ?t` until ?t is solved, based on other information. 

However, instance resolution still involves solving some metas, but only metas which are specifically created and marked for the purpose of instance head matching. For example, if we have `instance Foo (a, Bool)`, and the constraint `Foo (Bool, Bool)`, then a convenient implementation of matching is to create fresh `?a`, and solve `(?a, Bool) ~ (Bool, Bool)`. However, ?a is not a normal meta and ~ is not just usual unification, beacuse ~ can only solve instance head metas like ?a.

**Comment by atennapel** [comment]
On an somewhat unrelated note, what do you think about generalisation?
For example: `\x. x` will be generalised to `{t : *} -> t -> t`.
It seems it's harder to do in a dependently typed system. I think even in Haskell I could live without it. The only point where it's nice is in the REPL.

**Comment by AndrasKovacs** [comment]
Generalisation is only possible efficiently with scoped metas, because only metas which are local to a definition can be generalized. In H-M setting, the scoping of metas is very simple, see e.g. the following:

http://okmij.org/ftp/ML/generalization.html

In TT, scoped metas are harder. AFAIK GHC already employs a system for scoped constraints which is similar, see the last third of the  following:

https://www.youtube.com/watch?v=x3evzO8O9e8

I think that generalization is useful to have at least in a distinguished top scope. I would skip local let generalization personally to simplify things.

**Comment by atennapel** [comment]
I have implemented some type inference algorithms for System F and there top level generalization is really easy. You just gather the unsolved metas of the type and put them as foralls in front of the type. The only part you have to think about is the order of the foralls, which you probably want to be in order of occurrence in the type. Will that approach also work for dependent types?

**Comment by AndrasKovacs** [comment]
If you do that, you have to make sure that every meta in your program so far is solved when you start to elaborate the next top-level definition, or do what Agda does and disallow solving any meta from a previous top-level binding group.

Otherwise, it is possible that a meta in a previous top-level definition gets solved with an unsolved meta from the currently elaborated definition. Then it's not possible anymore to generalize over that meta.
